File: react-native/react-native-clockin-vision/ios/Face/Proximity.swift
//
//  Proximity.swift
//  TOTVSClockInVision
//
//  Created by Jansel Rodriguez on 8/1/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation
import TOTVSCameraKit

/**
* Proximity Detector
*/
open class Proximity : VisionReceiver<FaceObject> { }

/**
 * Data representing a proximity detection.
 */
public struct ProximityResult {
    let threshold: Float
    let faceWidth: Float
    let faceHeight: Float
    let isUnderThreshold: Bool
    
    public init(threshold: Float, faceWidth: Float, faceHeight: Float, isUnderThreshold: Bool = false) {
        self.threshold = threshold
        self.faceWidth = faceWidth
        self.faceHeight = faceHeight
        self.isUnderThreshold = isUnderThreshold
    }
}

/**
* Proximity Detector that act on [FaceObject] width dimensions.
*/
open class ProximityByFaceWidth : Proximity {
    /**
     * Value that limit the face width to determine when the face width is right, hence
     * regarded as close.
     */
    var threshold: Float = 100.0 {
        didSet { }
    }
    
    open var onProximity: ((ProximityResult) -> Void)? = nil
    
    /**
     * Control whether we've found a face or not.
     *
     * We use this for situations when the last event we emitted was a
     * valid event for either a right or wrong face and then no face is found.
     * We want to clean that previous state by emitting one event indicating that we're
     * not matching (finding) faces anymore. This allows us to send only one event of this type
     * since it might be the one with the most occurrence.
     *
     * We start with the assumption that we have found a face and let's the
     * detection disprove that assumption
     */
    private var foundFace = true
    
    public override init() {
    }
    
    public init(threshold: Float, onProximity: @escaping (ProximityResult) -> Void) {
        self.threshold = threshold
        self.onProximity = onProximity
    }
    
    open override func send(value: FaceObject) {
        if value == NullFaceObject {
            if foundFace {
                foundFace = false
                onProximity?(ProximityResult(
                    threshold: threshold,
                    faceWidth: 0.0,
                    faceHeight: 0.0
                ))
            }
        } else {
            foundFace = true
            // send a valid event.
            onProximity?(ProximityResult(
                threshold: threshold,
                faceWidth: Float(value.width),
                faceHeight: Float(value.height),
                isUnderThreshold: Float(value.width) <= threshold
            ))
        }
    }
}


extension ProximityByFaceWidth : CustomStringConvertible {
    public var description: String { "ProximityByFaceWidth" }
}

File: react-native/react-native-clockin-vision/android/src/main/java/com/totvs/clockin/vision/face/Proximity.kt
package com.totvs.clockin.vision.face

import android.graphics.Rect
import android.graphics.RectF
import android.util.Log
import android.util.Size
import androidx.core.graphics.toRect
import com.totvs.camera.view.CameraView
import com.totvs.camera.view.GraphicOverlay
import com.totvs.camera.vision.face.FaceObject
import com.totvs.camera.vision.face.NullFaceObject
import com.totvs.camera.vision.stream.VisionReceiver
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Proximity Detector
 */
interface Proximity : VisionReceiver<FaceObject>

/**
 * Proximity Detector that act on [FaceObject] width dimensions.
 */
open class ProximityByFaceWidth(
    private val onProximity: (ProximityResult) -> Unit
) : Proximity {
    /**
     * Value that limit the face width to determine when the face width is right, hence
     * regarded as close.
     */
    var threshold: Float = 100f
        set(value) = synchronized(this) {
            field = value
        }
        get() = synchronized(this) {
            field
        }

    /**
     * Control whether we've found a face or not.
     *
     * We use this for situations when the last event we emitted was a
     * valid event for either a right or wrong face and then no face is found.
     * We want to clean that previous state by emitting one event indicating that we're
     * not matching (finding) faces anymore. This allows us to send only one event of this type
     * since it might be the one with the most occurrence.
     *
     * We start with the assumption that we have found a face and let's the
     * detection disprove that assumption
     */
    protected val foundFace = AtomicBoolean(true)

    override fun send(value: FaceObject) {
        if (NullFaceObject == value) {
            if (foundFace.compareAndSet(true, false)) {
                onProximity(ProximityResult(
                    threshold = threshold,
                    faceWidth = 0f,
                    faceHeight = 0f
                ))
            }
        } else {
            foundFace.set(true)
            // send a valid event.
            onProximity(ProximityResult(
                isUnderThreshold = value.width < threshold,
                threshold = threshold,
                faceWidth = value.width,
                faceHeight = value.height
            ))
        }
    }

    override fun toString(): String = TAG

    /**
     * Data representing a proximity detection.
     */
    data class ProximityResult(
        val threshold: Float,
        val faceWidth: Float,
        val faceHeight: Float,
        val isUnderThreshold: Boolean = false
    )

    companion object {
        private const val TAG = "ProximityByFaceWidth"
    }
}

/**
 * Proximity Detector that act on [FaceObject] width dimensions plus assessing that face is well
 * located in camera view frame.
 */
class ProximityByFaceWidthConstrained(
    private val onProximity: (ProximityResult) -> Unit
) : ProximityByFaceWidth(onProximity) {

    override fun send(value: FaceObject) {
        if (NullFaceObject == value) {
            if (foundFace.compareAndSet(true, false)) {
                onProximity(ProximityResult(
                    threshold = threshold,
                    faceWidth = 0f,
                    faceHeight = 0f
                ))
            }
        } else {
            foundFace.set(true)

            val isConstrainedInSource = value.boundingBox?.let { faceBounds ->
                value.sourceSize.toRect().contains(faceBounds.toRect())
            } ?: true
            // send a valid event.
            onProximity(ProximityResult(
                isUnderThreshold = value.width < threshold || !isConstrainedInSource,
                threshold = threshold,
                faceWidth = value.width,
                faceHeight = value.height
            ))
        }
    }

    private fun Size.toRect() = Rect(0, 0, width, height)

    override fun toString(): String = TAG

    companion object {
        private const val TAG = "ProximityByFaceWidthConstrained"
    }
}
